### PR TITLE
ICS Entity Index

### LINKED ISSUE #20: ICS Entity Index
Write an entity index for finding components by entity ID

### PR DESCRIPTION
**Pull request by joeltio**
Closes #20.

Adds a set intersection utility and the entity index with tests.

### CODE REVIEW

File: sim/src/index/entity.test.cpp [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    index.addComponent(entityId2, std::pair(3, 8));
+
+    ASSERT_EQ(index.getComponents(entityId1).size(), 1);
+    ASSERT_EQ(index.getComponents(entityId2).size(), 2);
+}
+
+TEST(TEST_SUITE, UseIndexIdFilter) {
+    Entity index;
+    auto entityId = index.addEntityId();
+    auto filter = index.id(entityId);
```
Comment (mrzzy) [inline sim/src/index/entity.test.cpp:45]: Even with the test its really hard to follow what is going on with this `.id()` method. 
After digging abit, I think this means that of the component store Ids that is attached to the entity and return the ids.

File: sim/include/index/entity.h [hunk 2]
```diff
@@ -0,0 +1,48 @@
+#ifndef BENTOBOX_ENTITY_H
+#define BENTOBOX_ENTITY_H
+
+#include <unordered_set>
+#include <core/ics/componentStore.h>
+#include <core/ics/util/setIntersection.h>
+
+namespace ics::index {
+
+    class Entity {
```
Comment (mrzzy) [inline sim/include/index/entity.h:10]: Naming this `Entity` makes think that this represents one Entity instead of a index of all Entities in the System.

**Review by mrzzy — COMMENTED** [review]
Code wise this looks mostly Ok. :+1: for short PR.

From the user perspective (probably me when implementing the gRPC service), having multiple layers of indirection is  kind of hard to keep in mind:
1. Entity -> CompStoreIds
2. CompStoreId -> ComponentStore
3. CompStoreIds stored in CompSets.
4.  CompGroup -> `Vector<Component>`

so I would appreciate if you added top-level function like  `ics::addComponent()` to provide a high-level API so that I don't to worry about this. Is that coming in another PR?

**Comment by joeltio** [comment]
There should be only 1 layer I believe:
CompStoreId -> Component (in ComponentStore)

If you want, you could always search through the whole component store for the CompStoreId that you want. All other classes or structs are made for further efficiency. Similar to how databases work, indexes only speed things up.

However, in our case, unlike databases, it doesn't use or update all indexes for you automatically. All indexes currently work on the basis that you have an extra piece of information that can speed up the query: the component type or the entity ID.

**Comment by mrzzy** [comment]
> However, in our case, unlike databases, it doesn't use or update all indexes for you automatically. All indexes currently work on the basis that you have an extra piece of information that can speed up the query: the component type or the entity ID.

Exactly why this could be difficult to  use correctly. I didn't write the code and could miss updating one index or another and then be confused on why things don't work. If these index updates are abstracted away this could be much easier.

I guess we can worry about this in a next PR.
 
Names updates makes things clearer :+1:
